zip/ditto_reader.go

```diff
@@ -0,0 +1,93 @@
+package zip
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+
+	"github.com/bitrise-io/go-utils/v2/log"
+
+	"github.com/bitrise-io/go-utils/v2/command"
+	"github.com/bitrise-io/go-utils/v2/env"
+	"github.com/bitrise-io/go-utils/v2/pathutil"
+)
+
+type dittoReader struct {
+	extractedDir string
+	logger       log.Logger
+}
+
+// IsDittoReaderAvailable ...
+func IsDittoReaderAvailable() bool {
+	_, err := exec.LookPath("ditto")
+	return err == nil
+}
+
+// NewDittoReader ...
+func NewDittoReader(archivePath string, logger log.Logger) (ReadCloser, error) {
+	factory := command.NewFactory(env.NewRepository())
+	tmpDir, err := pathutil.NewPathProvider().CreateTempDir("ditto_reader")
+	if err != nil {
+		return nil, err
+	}
+
+	/*
+	   -x            Extract the archives given as source arguments. The format
+	                 is assumed to be CPIO, unless -k is given.  Compressed CPIO
+	                 is automatically handled.
+
+	   -k            Create or extract from a PKZip archive instead of the
+	                 default CPIO.  PKZip archives should be stored in filenames
+	                 ending in .zip.
+	*/
+	cmd := factory.Create("ditto", []string{"-x", "-k", archivePath, tmpDir}, nil)
```

Comment:
There is a difference in the zip and ditto reader's behaviour. The ditto reader extracts the zip during initialisation while (if I am reading the zip reader's code right) it only starts extracting when you want to open a specific file from the archive. 

If we do the extraction right in the init phase then the init is not lightweight and has this hidden side effect. Which might not be a problem but it is a bit hidden. 

If you ask me right now I would do the extraction as part of an action. So in this case when the `ReadFile` method is called.